### PR TITLE
forkable zkbridge

### PR DESCRIPTION
**Pull request by josojo**
This PR is unfortunately much bigger than expected. If I should split for an easier review, let me know.

The implementation allows users to only use the bridge in the newest chain (youngest bridge child). This enables us to monitor the used deposit leaves effectively, by just looking up whether the leaf has been used in any parent contract (instead of searching in two directions). But this implies that the whole state of the depositTree(_branches originally called) need to be copied by each fork.

I tried to test it very well, with over 700 lines of test. But only with e2e tests, I will have full confidence.

### CODE REVIEW

File: development/contracts/ForkableBridge.sol [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        _setupRole(HARD_ASSET_MANAGER_ROLE, hardAssetManger);
+    }
+
+    /**
+     * @notice Function to send hard assets to the children-bridge contracts
+     * @param token Address of the token
+     * @param amount Amount of tokens to transfer
+     * @param to Address to transfer the tokens to
+     */
+    function manageHardAssets(
```
Comment (edmundedgar) [inline development/contracts/ForkableBridge.sol:48]: "manage" isn't very descriptive, can we express what this is for? is it something like transferHardAssetsToChild?

**Comment by edmundedgar** [comment]
On copying depositTree, it sounds bad to have what sounds like an unbounded amount of state we have to copy. IIUC it's just an ordered list of deposits requests? (If it's quicker to do it this was for now then it's ok to change it later)

Could we do something like

parent: depositTree

child: 
 - parentDepositTreeLastIndex
 - parentDepositTreeDeletions[idx] => true (if we need deletions)
 - depositTree
 ...then when the child needs to read depositTree it checks if the index is below parentDepositTreeLastIndex, if so check if it's deleted, and otherwise reads it off the parent?

**Comment by edmundedgar** [comment]
No need to split the PRs up. I'm finding the little PRs pretty hard to follow when we're still in the state of putting the whole thing together, they need to much context to stuff you're thinking of doing in different PRs. Once we get a basically functioning thing I think little PRs will make sense because each one is a small tweak that you can think about in isolation.

**Comment by josojo** [comment]
> On copying depositTree, it sounds bad to have what sounds like an unbounded amount of state we have to copy. IIUC it's just an ordered list of deposits requests? (If it's quicker to do it this was for now then it's ok to change it later)
> 
> Could we do something like
> 
> parent: depositTree
> 
> child:
> 
> * parentDepositTreeLastIndex
> * parentDepositTreeDeletions[idx] => true (if we need deletions)
> * depositTree
>   ...then when the child needs to read depositTree it checks if the index is below parentDepositTreeLastIndex, if so check if it's deleted, and otherwise reads it off the parent?

The depositTree (I will rename it to depositTreeHashes) are only 32 bytes32. I think its not unbounded, its only 30*32 gas. Its a hashtreeing tree that is different from the normal merkle tree. It requires less hashing, but also only one storage update. 
Take a look here: https://github.com/RealityETH/zkevm-contracts/blob/main/contracts/lib/DepositContract.sol its pretty nice. The advantage of current approach is that we need to change less zkevm-node code, as your proposal would change the way the proof is calculated.

**Review by edmundedgar — APPROVED** [review]
